tests/unit/changeset-test.js

```diff
@@ -1057,6 +1060,128 @@ module('Unit | Utility | changeset', function(hooks) {
+        return Promise.resolve(this);
+      },
+    });
+
+    let dummyChangeset = Changeset(mockUserModel);
+
+    assert.equal(get(dummyChangeset, 'isPristine'), true, 'changeset is pristine');
+
+    dummyChangeset.set('profile.firstName', 'Zoe');
+
+    assert.equal(get(dummyChangeset, 'isPristine'), false, 'changeset is not pristine as there is a change');
+
+    assert.equal(dummyChangeset.get('pendingData.profile.firstName'), 'Zoe', 'Model belongsTo property should be updated')
+    assert.equal(dummyChangeset.get('pendingData.profile.lastName'), 'Ross', 'Existing property should stay the same')
+
+    assert.equal(dummyChangeset.get('data.profile.firstName'), 'Bob', 'Original model property should stay without changes')
+    assert.equal(dummyChangeset.get('data.profile.lastName'), 'Ross', 'Original model property should stay without changes')
+  });
+
+
+  todo('#pendingChanges with ember-data model with multiple change steps', async function(assert) {
+    let store = this.owner.lookup('service:store');
+
+    let mockProfileModel = store.createRecord('profile');
+    let mockUserModel = store.createRecord('user', {
+      profile: mockProfileModel,
+      save: function() {
+        return Promise.resolve(this);
+      },
+    });
+
+    let dummyChangeset = Changeset(mockUserModel);
+    assert.equal(get(dummyChangeset, 'isPristine'), true, 'changeset is pristine');
+
+    dummyChangeset.set('profile.firstName', 'Zoe');
+
+    assert.equal(get(mockUserModel, 'profile.firstName'), 'Bob', 'belongsTo property is not modified');
+    assert.equal(get(dummyChangeset, 'profile.firstName'), 'Zoe', 'belongsTo changeset is modified');
+    assert.equal(get(dummyChangeset.pendingData, 'profile.firstName'), 'Zoe', 'pending data should give updated values');
+
+    await dummyChangeset.save();
+
+    assert.equal(get(mockUserModel, 'profile.firstName'), 'Zoe', 'belongsTo property is updated after save');
+    assert.equal(get(dummyChangeset, 'profile.firstName'), 'Zoe', 'belongsTo changeset has new value');
+    assert.equal(get(dummyChangeset.pendingData, 'profile.firstName'), 'Zoe', 'pendingData should have a new value');
+
+    let updatedMockProfileModel = store.createRecord('profile');
+
+    dummyChangeset.set('profile', updatedMockProfileModel);
+
+    assert.equal(get(mockUserModel, 'profile.firstName'), 'Zoe', 'belongsTo property is model has initial value');
+    assert.equal(get(dummyChangeset, 'profile.firstName'), 'Bob', 'belongsTo changeset is updated');
+    assert.equal(get(dummyChangeset.pendingData, 'profile.firstName'), 'Bob', 'belongsTo pendingData has new value');
+
+    await dummyChangeset.save();
+
+    assert.equal(get(mockUserModel, 'profile.firstName'), 'Bob', 'belongsTo property is updated');
+    assert.equal(get(dummyChangeset, 'profile.firstName'), 'Bob', 'belongsTo changeset is updated');
+    assert.equal(get(dummyChangeset.pendingData, 'profile.firstName'), 'Bob', 'belongsTo pendingData is updated');
+  });
```

Comment:
What isn't working now with e-d objects?  If you hit the circular dependency issue, wouldn't we have an `assert.throws` assertion?

Reply:
circular dependency issue happens in one test, but not in another though. And `assert.throws` didn't work in that case. 2nd tests was throwing exception correctly. I'll test it out more.

Reply:
Weird, I wasn't able to reproduce it. I think what I was seeing:

I've added console.error() in that new try/catch block, to see the error. Then I've added logging in mergeDeep, to see which keys are failing. And then when I saw that it fails with `null`, I've added `possible !== null` and after that it was just going in loops, without breaking through max call stack. But I think it was my failed experiments :) Now everything seems to be fine.